lib/sandbox.jail.ml

```diff
@@ -0,0 +1,157 @@
+      | args -> List.rev (List.rev_map Filename.quote args)
+    in
+    let commandline =
+      match env with
+      | [] -> commandline
+      | _ -> "env" :: List.rev_append env commandline
+    in
+    let commandline =
+      String.concat " "
+        ([ "cd" ; Filename.quote config.cwd ; "&&" ] @ commandline)
+    in
+    (* Ask for a login shell in order to properly source opam settings. *)
+    [ "command=/usr/bin/su" ; "-l" ; username ; "-c" ; commandline ]
+  in
+  let jailname = "name=obuilder" in
+  let path = "path=" ^ rootdir in
+  let devfs_setup = "mount.devfs" in
+  let options =
+    let options = [ jailname ; path ; devfs_setup ] in
+    match config.network with
+    | [ "host" ] ->
+      "ip4=inherit" :: "ip6=inherit" :: "host=inherit" :: options
+    | _ -> options
+  in
+  List.rev_append options commandline
+
+let copy_to_log ~src ~dst =
+  let buf = Bytes.create 4096 in
+  let rec aux () =
+    Lwt_unix.read src buf 0 (Bytes.length buf) >>= function
+    | 0 -> Lwt.return_unit
+    | n -> Build_log.write dst (Bytes.sub_string buf 0 n) >>= aux
+  in
+  aux ()
+
+let run ~cancelled ?stdin:stdin ~log (t : t) config rootdir =
+  let cwd = rootdir in
+  Lwt_mutex.with_lock t.lock (fun () ->
+  Os.with_pipe_from_child @@ fun ~r:out_r ~w:out_w ->
+  let rootdir = rootdir / "rootfs" in
+  let workdir = rootdir / config.Config.cwd in
+  (* Make sure the work directory exists prior to starting the jail. *)
+  begin
+    match Os.check_dir workdir with
+    | `Present -> Lwt.return_unit
+    | `Missing -> Os.sudo [ "mkdir" ; "-p" ; workdir ]
+  end >>= fun () ->
+  let stdout = `FD_move_safely out_w in
+  let stderr = stdout in
+  let copy_log = copy_to_log ~src:out_r ~dst:log in
+  let proc =
+    let cmd =
+      let options = jail_options config rootdir in
+      "jail" :: "-c" :: options
+    in
+    let stdin = Option.map (fun x -> `FD_move_safely x) stdin in
+    let pp f = Os.pp_cmd f ("", cmd) in
+    (* This is similar to
+       Os.sudo_result ~cwd ?stdin ~stdout ~stderr ~pp cmd
+       but also unmounting the in-jail devfs if necessary, see below. *)
+    let cmd = if Os.running_as_root then cmd else "sudo" :: "--" :: cmd in
+    Logs.info (fun f -> f "Exec %a" Os.pp_cmd ("", cmd));
+    !Os.lwt_process_exec ~cwd ?stdin ~stdout ~stderr ~pp
+      ("", Array.of_list cmd) >>= function
+    | Ok 0 ->
+      (* If the command within the jail completes, the jail is automatically
+         removed, but without performing any of the stop and release actions,
+         thus we can not use "exec.stop" to unmount the in-jail devfs
+         filesystem. Do this here, ignoring the exit code of umount(8). *)
+      let cmd = [ "sudo" ; "/sbin/umount" ; rootdir / "dev" ] in
+      Os.exec ~is_success:(fun _ -> true) cmd >>= fun () ->
+      Lwt_result.ok Lwt.return_unit
+    | Ok n -> Lwt.return @@ Fmt.error_msg "%t failed with exit status %d" pp n
+    | Error e -> Lwt_result.fail (e: [`Msg of string] :> [> `Msg of string])
```

Comment:
Is the coercion really needed?

Reply:
It has been copied from `exec_result` in `lib/os.ml` so I suppose it is needed...

Reply:
...but it is indeed not necessary.